Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py

```diff
@@ -0,0 +1,252 @@
+        raise ConfigException("A kubeconfig needs to be set in order to test a k8s cluster version.")
+
+    return config
+
+
+async def get_k8s_cluster_labelled_nodes(kubeconfig):
+    """Get the labels of the k8s cluster under test."""
+    _ = await kubernetes_asyncio.config.load_kube_config(kubeconfig)
+
+    nodes = list()
+    labels = [
+        "kubernetes.io/hostname",
+        "topology.kubernetes.io/zone",
+        "topology.kubernetes.io/region",
+        "node-role.kubernetes.io/control-plane"
+    ]
+
+    async with kubernetes_asyncio.client.ApiClient() as api:
+        core_api = kubernetes_asyncio.client.CoreV1Api(api)
+        result = await core_api.list_node()
+        for item in result.items:
+            node = dict()
+            node_labels = item.metadata.labels
+            for label in labels:
+                try:
+                    node[label] = node_labels[label]
+                except KeyError:
+                    # Node is not labelled as a control plane node
+                    pass
+            nodes.append(node)
+
+    return nodes
+
+
+def compare_labels(node_list, node_type="master"):
+
+    labels = {
+        "topology.kubernetes.io/region": set(),
+        "topology.kubernetes.io/zone": set(),
+        "kubernetes.io/hostname": set(),
+    }
+
+    for node in node_list:
+        for key in labels:
+            labels[key].add(node[key])
+
+    for label in labels:
+        if len(labels[label]) <= 1:
+            if len(labels[label]) == 1:
+                logger.warning(f"There seems to be no distribution across multiple {label.split('/')[1]}s "
+                               "or labels aren't set correctly across nodes.")
+            elif len(labels[label]) == 0:
+                logger.warning(f"The label for {label.split('/')[1]}s don't seem to be set for all nodes.")
+        else:
+            logger.info(f"The nodes are distributed across {str(len(labels[label]))} {label.split('/')[1]}s.")
+            return
+
+    if node_type == "master":
+        raise DistributionException()
+    elif node_type == "worker":
+        logger.warning("No node distribution could be detected for the worker nodes. "
+                       "This produces only a warning, since it is just a recommendation.")
+        return
+
+
+async def main(argv):
+    try:
+        config = initialize_config(parse_arguments(argv))
+    except (OSError, ConfigException, HelpException) as e:
+        if hasattr(e, 'message'):
+            logger.error(e.message)
+        print_usage()
+        return 1
+
+    nodes = await get_k8s_cluster_labelled_nodes(config.kubeconfig)
+
+    if len(nodes) < 2:
+        logger.error("The tested cluster only contains a single node, which can't comply with the standard.")
+        return 2
+
+    labelled_master_nodes = [node for node in nodes if "node-role.kubernetes.io/control-plane" in node]
+    try:
+        if len(labelled_master_nodes) > 1 and len(nodes) > 2:
```

Comment:
shouldn't this be `>= 1` and `=> 2`?

Reply:
No, since if there is only 1 master node, we can't compare it with another master node. And we can also only do this, if we have more than 2 nodes in general, since this would mean we have at least one worker node.

I could argue about `> 2` being `>= 2`, since there could be a cluster without worker nodes and we would do these edge cases in different `if else` paths right now.

Reply:
My understanding was that once there is at least one occurrence of "node-role.kubernetes.io/control-plane", then we know that these labels are set, and we can distinguish between master nodes and worker nodes; otherwise we just assume that every node is a master node. Can a cluster with only one master node pass the test at all?

Reply:
> Can a cluster with only one master node pass the test at all?
I would personally say no, but not because the setup is wrong, but because such a cluster wouldn't be constructed for availability or long-lasting services, but its probably there for development, testing, etc.
So its probably not something you would test for that.

Reply:
My question was meant with respect to the standard, not your opinion ;)

Reply:
If you change the condition to `>= 1`, then such a setup would fail the test (because `compare_labels` wouldn't reach the return statement, but the raise statement), and if the standard agrees with that result, then I don't see why we need `> 1`.

Reply:
The standard says that multiple control plane nodes should be distributed. So it would pass.

Reply:
Got it.